clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -111,6 +124,24 @@ public synchronized Cluster fetch() {
         return cache.cluster();
     }
 
+    /**
+     * Increment the nodesTriedSinceLastSuccessfulRefresh
+     */
+    public synchronized void incrementNodesTriedSinceLastSuccessfulRefresh() {
+        this.nodesTriedSinceLastSuccessfulRefresh++;
+    }
+
+    /**
+     * Whether the client should update the cluster metadata
+     * @param nowMs
+     * @return true if client hasn't refreshed cluster metadata for maxClusterMetadataExpireTimeMs and
+     * has tried connecting to at least half of the nodes in current node set
+     */
+    public synchronized boolean shouldUpdateClusterMetadata(long nowMs) {
+        return this.nodesTriedSinceLastSuccessfulRefresh >= this.cache.cluster().nodes().size() / 2 &&
```

Comment:
instead of trying half of the nodes, maybe it is sufficient to update cluster metadata from bootstrap server after the first attempt, since it is rare to not update metadata for maxClusterMetadataExpireTimeMs. We want a quick way out of this.  (each attempt would potentially need to wait for a connection timeout).

Reply:
with default values, connection timeout->retrying happens at 50ms interval, so in a cluster of size ~80 in venice case, it's ~40*50ms = 2s, so it's not that long. But change it to 1 anyway since this should be rare.